File: src/backend/decoders/mndp.py
"""MNDP (MikroTik Neighbor Discovery Protocol) decoder for PacketSnitch backend.

Decodes MNDP frames carried over UDP port 5678 (broadcast). MNDP is a
MikroTik Layer 2/3 protocol used to share device information between
directly connected RouterOS equipment.

MNDP header (4 bytes):
  - Unknown / Preamble (2 bytes — observed as 0x0000 in captures)
  - Sequence Number    (2 bytes, BE)
  - TLVs follow

MNDP TLV structure (differs from CDP — length is value-only):
  - Type           (2 bytes, BE)
  - Length         (2 bytes, BE — value length, NOT including Type+Length)
  - Value          (Length bytes)

This decoder follows the same dual-key dict pattern as the other decoders,
returning a dict on success or ``None`` on mismatch.
"""

import struct

MNDP_TLV_TYPES = {
    0x0001: "MAC Address",
    0x0005: "Identity",
    0x0007: "Version",
    0x0008: "Platform",
    0x000A: "Uptime",
    0x000B: "Software ID",
    0x000C: "Board",
    0x000E: "Unpack",
    0x000F: "IPv6 Address",
    0x0010: "Interface Name",
    0x0011: "IPv4 Address",
}

MNDP_UNPACK_VALUES = {
    1: "None",
}


def _formatMac(raw):
    if raw is None or len(raw) < 6:
        return "00:00:00:00:00:00"
    return ":".join(f"{b:02x}" for b in raw[:6])


def _formatIpv4(raw):
    if raw is None or len(raw) < 4:
        return "0.0.0.0"
    return ".".join(str(b) for b in raw[:4])


def _formatIpv6(raw):
    if raw is None or len(raw) < 16:
        return "::"
    parts = []
    for i in range(0, 16, 2):
        parts.append(f"{(raw[i] << 8) | raw[i + 1]:x}")
    return ":".join(parts)


def _formatUptime(raw):
    """Uptime TLV: 4-byte big-endian centiseconds (1/100 s)."""
    if not raw or len(raw) < 4:
        return None
    try:
        centis = struct.unpack_from(">I", raw, 0)[0]
    except Exception:
        return None
    seconds = centis / 100.0
    return f"{seconds:.2f}s"


def decodeMNDP(p, rawPayload):
    """Decode an MNDP frame from raw UDP payload bytes.

    Returns a dict on success or ``None`` when the bytes do not match
    the MNDP wire format.
    """
    try:
        if rawPayload is None or len(rawPayload) < 8:
            return None

        # MNDP header: 2 bytes unknown/preamble + 2 bytes sequence number.
        headerUnknown = rawPayload[0:2]
        seqNo = struct.unpack_from(">H", rawPayload, 2)[0]

        result = {
            "SeqNo": seqNo,
            "mndp.seqno": seqNo,
            "link.mndp.seqno": seqNo,
            "Header Unknown": headerUnknown.hex(),
            "mndp.header.unknown": headerUnknown.hex(),
            "link.mndp.header.unknown": headerUnknown.hex(),
            "Wire length": len(rawPayload),
            "wire.len": len(rawPayload),
            "link.mndp.wire.len": len(rawPayload),
        }

        # Parse TLVs starting after the 4-byte header. Unlike CDP, the
        # MNDP TLV length field is the value length only (it does NOT
        # include the 4-byte Type+Length header).
        tlvOff = 4
        tlvs = []
        firstTlvType = None
        while tlvOff + 4 <= len(rawPayload):
            tlvType = struct.unpack_from(">H", rawPayload, tlvOff)[0]
            tlvLen = struct.unpack_from(">H", rawPayload, tlvOff + 2)[0]
            if tlvOff + 4 + tlvLen > len(rawPayload):
                break
            tlvValue = rawPayload[tlvOff + 4: tlvOff + 4 + tlvLen]
            if firstTlvType is None:
                firstTlvType = tlvType
            tlvName = MNDP_TLV_TYPES.get(tlvType, f"Unknown (0x{tlvType:04x})")

            if tlvType == 0x0001:
                mac = _formatMac(tlvValue)
                result["MAC Address"] = mac
                result["mndp.mac"] = mac
                result["link.mndp.mac"] = mac
            elif tlvType == 0x0005:
                identity = tlvValue.decode(errors="ignore").rstrip("\x00")
                result["Identity"] = identity
                result["mndp.identity"] = identity
                result["link.mndp.identity"] = identity
            elif tlvType == 0x0007:
                version = tlvValue.decode(errors="ignore").rstrip("\x00")
                result["Version"] = version
                result["mndp.version"] = version
                result["link.mndp.version"] = version
            elif tlvType == 0x0008:
                platform = tlvValue.decode(errors="ignore").rstrip("\x00")
                result["Platform"] = platform
                result["mndp.platform"] = platform
                result["link.mndp.platform"] = platform
            elif tlvType == 0x000A:
                uptime = _formatUptime(tlvValue)
                if uptime is not None:
                    result["Uptime"] = uptime
                    result["mndp.uptime"] = uptime
                    result["link.mndp.uptime"] = uptime
            elif tlvType == 0x000B:
                softwareId = tlvValue.decode(errors="ignore").rstrip("\x00")
                result["Software ID"] = softwareId
                result["mndp.software_id"] = softwareId
                result["link.mndp.software_id"] = softwareId
            elif tlvType == 0x000C:
                board = tlvValue.decode(errors="ignore").rstrip("\x00")
                result["Board"] = board
                result["mndp.board"] = board
                result["link.mndp.board"] = board
            elif tlvType == 0x000E:
                if len(tlvValue) >= 1:
                    unpackVal = int(tlvValue[0])
                    unpackName = MNDP_UNPACK_VALUES.get(
                        unpackVal, f"Unknown ({unpackVal})"
                    )
                    result["Unpack"] = unpackName
                    result["mndp.unpack"] = unpackName
                    result["link.mndp.unpack"] = unpackName
            elif tlvType == 0x000F:
                ipv6 = _formatIpv6(tlvValue)
                result["IPv6 Address"] = ipv6
                result["mndp.ipv6_address"] = ipv6
                result["link.mndp.ipv6_address"] = ipv6
            elif tlvType == 0x0010:
                ifaceName = tlvValue.decode(errors="ignore").rstrip("\x00")
                result["Interface Name"] = ifaceName
                result["mndp.interface_name"] = ifaceName
                result["link.mndp.interface_name"] = ifaceName
            elif tlvType == 0x0011:
                ipv4 = _formatIpv4(tlvValue)
                result["IPv4 Address"] = ipv4
                result["mndp.ipv4_address"] = ipv4
                result["link.mndp.ipv4_address"] = ipv4

            tlvs.append({"type": tlvName, "length": tlvLen})
            tlvOff += 4 + tlvLen

        # Strict gate: the first TLV must be the MAC Address TLV (type 1)
        # and must have length 6. This mirrors Wireshark's heuristic and
        # prevents random binary data from being accepted as MNDP.
        if firstTlvType != 0x0001:
            return None

        if tlvs:
            result["TLVs"] = tlvs
            result["mndp.tlvs"] = tlvs
            result["link.mndp.tlvs"] = tlvs
        return result
    except Exception:
        return None

File: src/ui/decoders/conv/mndp.js
// MNDP Conv decoder: parses MikroTik Neighbor Discovery Protocol frames
// carried over UDP port 5678 (broadcast). Identifies messages by the
// 4-byte header (2 unknown + 2 seqno) and the mandatory first TLV type 1
// (MAC Address, length 6) heuristic, then decodes the TLV payload.
//
// Unlike CDP, the MNDP TLV length field is the value length only (it does
// NOT include the 4-byte Type+Length header).

const MNDP_TLV_TYPES = {
    0x0001: "MAC Address",
    0x0005: "Identity",
    0x0007: "Version",
    0x0008: "Platform",
    0x000a: "Uptime",
    0x000b: "Software ID",
    0x000c: "Board",
    0x000e: "Unpack",
    0x000f: "IPv6 Address",
    0x0010: "Interface Name",
    0x0011: "IPv4 Address",
};

const MNDP_UNPACK_VALUES = {
    1: "None",
};

function readUint16BE(bytes, offset) {
    return (bytes[offset] << 8) | bytes[offset + 1];
}

function readUint32BE(bytes, offset) {
    return (
        ((bytes[offset] << 24) |
            (bytes[offset + 1] << 16) |
            (bytes[offset + 2] << 8) |
            bytes[offset + 3]) >>>
        0
    );
}

function formatMac(bytes, offset) {
    const parts = [];
    for (let i = 0; i < 6; i++) {
        parts.push(bytes[offset + i].toString(16).padStart(2, "0"));
    }
    return parts.join(":");
}

function formatIpv4(bytes, offset) {
    return `${bytes[offset]}.${bytes[offset + 1]}.${bytes[offset + 2]}.${bytes[offset + 3]}`;
}

function formatIpv6(bytes, offset) {
    const parts = [];
    for (let i = 0; i < 16; i += 2) {
        parts.push(((bytes[offset + i] << 8) | bytes[offset + i + 1]).toString(16));
    }
    return parts.join(":");
}

function formatUptime(value) {
    // Uptime TLV: 4-byte big-endian centiseconds (1/100 s)
    if (value.length < 4) return null;
    const centis = readUint32BE(value, 0);
    const seconds = centis / 100;
    return `${seconds.toFixed(2)}s`;
}

function decodeMndpFromBytes(bytes) {
    if (!(bytes instanceof Uint8Array) || bytes.length < 8) return null;

    // MNDP header: 2 bytes unknown/preamble + 2 bytes sequence number (BE)
    const seqNo = readUint16BE(bytes, 2);

    const fields = [
        { name: "SeqNo", value: seqNo },
        { name: "Header Unknown", value: Array.from(bytes.slice(0, 2)).map((b) => b.toString(16).padStart(2, "0")).join("") },
    ];

    // Parse TLVs. MNDP TLV length is value-only (excludes the 4-byte T+L).
    let tlvOff = 4;
    let tlvCount = 0;
    let firstTlvType = null;
    while (tlvOff + 4 <= bytes.length) {
        const tlvType = readUint16BE(bytes, tlvOff);
        const tlvLen = readUint16BE(bytes, tlvOff + 2);
        if (tlvOff + 4 + tlvLen > bytes.length) break;
        const tlvValue = bytes.slice(tlvOff + 4, tlvOff + 4 + tlvLen);
        if (firstTlvType === null) firstTlvType = tlvType;
        const tlvName = MNDP_TLV_TYPES[tlvType] || `Unknown (0x${tlvType.toString(16).padStart(4, "0")})`;

        if (tlvType === 0x0001) {
            if (tlvValue.length >= 6) {
                fields.push({ name: "MAC Address", value: formatMac(tlvValue, 0) });
            }
        } else if (tlvType === 0x0005) {
            fields.push({ name: "Identity", value: new TextDecoder().decode(tlvValue).replace(/\x00+$/, "") });
        } else if (tlvType === 0x0007) {
            fields.push({ name: "Version", value: new TextDecoder().decode(tlvValue).replace(/\x00+$/, "") });
        } else if (tlvType === 0x0008) {
            fields.push({ name: "Platform", value: new TextDecoder().decode(tlvValue).replace(/\x00+$/, "") });
        } else if (tlvType === 0x000a) {
            const uptime = formatUptime(tlvValue);
            if (uptime) fields.push({ name: "Uptime", value: uptime });
        } else if (tlvType === 0x000b) {
            fields.push({ name: "Software ID", value: new TextDecoder().decode(tlvValue).replace(/\x00+$/, "") });
        } else if (tlvType === 0x000c) {
            fields.push({ name: "Board", value: new TextDecoder().decode(tlvValue).replace(/\x00+$/, "") });
        } else if (tlvType === 0x000e && tlvValue.length >= 1) {
            const unpackVal = tlvValue[0];
            const unpackName = MNDP_UNPACK_VALUES[unpackVal] || `Unknown (${unpackVal})`;
            fields.push({ name: "Unpack", value: unpackName });
        } else if (tlvType === 0x000f) {
            if (tlvValue.length >= 16) {
                fields.push({ name: "IPv6 Address", value: formatIpv6(tlvValue, 0) });
            }
        } else if (tlvType === 0x0010) {
            fields.push({ name: "Interface Name", value: new TextDecoder().decode(tlvValue).replace(/\x00+$/, "") });
        } else if (tlvType === 0x0011) {
            if (tlvValue.length >= 4) {
                fields.push({ name: "IPv4 Address", value: formatIpv4(tlvValue, 0) });
            }
        }
        tlvOff += 4 + tlvLen;
        tlvCount += 1;
    }

    // Strict gate (mirrors Wireshark's heuristic): the first TLV must be
    // the MAC Address TLV (type 1) with length 6. This prevents random
    // binary data from being accepted as MNDP.
    if (firstTlvType !== 0x0001) return null;

    // A real MNDP frame must have at least one TLV.
    if (tlvCount === 0) return null;

    return { protocol: "MNDP", fields };
}

module.exports = { decodeMndpFromBytes };